Make Solid Cache the default for new Rails apps

djmb opened the pull request:
The defaults for the cache are pulled in from the [template](https://github.com/rails/solid_cache/blob/87fd0e9611237b5608b5e0f4197aec74fe333ca4/lib/generators/solid_cache/install/templates/config/solid_cache.yml.tt) in Solid Cache. Currently 1 week max age and 256MB max size.

Right now this sets up Solid Cache to be used in development and production - maybe we'd want to stick with a memory cache in development?

When using Sqlite, we'll use a separate database for the cache, with the migrations for that db going into `db/cache/migrate`. We then need to adjust the db change task to move the migrations if required.

For all databases we'll use a separate connection pool for the database even if it points to the primary database. This stops cache queries from being affected by primary database transactions.

As a result the `config/database.yml` templates have been converted to the three-tier format. We could keep the two-tier format when not using Solid Cache, but the templates would start to get messy.

Resolves #50443 cc @dhh

djmb committed:
Make Solid Cache the default for new Rails apps

The defaults for the cache are pulled in from the [template](https://github.com/rails/solid_cache/blob/87fd0e9611237b5608b5e0f4197aec74fe333ca4/lib/generators/solid_cache/install/templates/config/solid_cache.yml.tt)
in Solid Cache. Currently 1 week max age and 256MB max size.

Right now this sets up Solid Cache to be used in development and
production - maybe we'd want to stick with a memory cache in development?

When using Sqlite, we'll use a separate database for the cache, with
the migrations for that db going into `db/cache/migrate`. We then need
to adjust the db change task to move the migrations if required.

For all databases we'll use a separate connection pool for the database
even if it points to the primary database. This stops cache queries
from being affected by primary database transactions.

As a result the `config/database.yml` templates have been converted to
the three-tier format. We could keep the two-tier format when not using
Solid Cache, but the templates would start to get messy.

dhh commented:
Excellent! After thinking about this more, I think the right move for both Solid Queue and Cache is not to run them in development. We haven't done that with any of the Resque options they're replacing. So let's just do the configuration for production only for now.

dhh commented:
Also, for production, I would bump the limits substantially. Also, I think we should bump the time limit to 30 days, rather than just one week, for the default. Wish we could also bump the space, but on Heroku's most stingy plan, you only get 1GB(!!).

dhh commented on railties/lib/rails/generators/rails/app/templates/config/environments/production.rb.tt:
```diff
@@ -73,8 +73,13 @@ Rails.application.configure do
   # config.cache_store = :mem_cache_store
+  <%- else -%>
+  # Cache with Solid Cache
+  config.cache_store = :solid_cache_store
```
We should expose the limit settings here. Both the time and the space one.

dhh replied:
Nevermind, they're exposed in the dedicated config file 👍

djmb committed:
Test fixes

djmb committed:
Fix generation tests to find migrations correctly

djmb committed:
Don't use Solid Cache in development

djmb committed:
More test fixes

djmb committed:
Exclude migration fixture from rubocop

packagethief commented on railties/lib/rails/generators/app_base.rb:
```diff
@@ -640,6 +648,13 @@ def cable_gemfile_entry
+      def solid_cache_gemfile_entry
+        return if options[:skip_solid_cache]
+
+        comment = "Use Solid Cache for caching (http://github.com/rails/solid_cache)"
```
```suggestion
        comment = "Use Solid Cache for caching (https://github.com/rails/solid_cache)"
```

packagethief commented on railties/lib/rails/generators/app_base.rb:
```diff
@@ -640,6 +648,13 @@ def cable_gemfile_entry
       end
 
+      def solid_cache_gemfile_entry
+        return if options[:skip_solid_cache]
```
Maybe use the `skip_solid_cache?` method instead?

packagethief commented on railties/lib/rails/generators/rails/app/templates/config/databases/mysql.yml.tt:
```diff
@@ -22,15 +22,25 @@ default: &default
+  primary:<% unless options[:skip_solid_cache] %> &development_primary<% end %>
+    <<: *default
+    database: <%= app_name %>_development
+<% unless options[:skip_solid_cache] -%>
```
I'm not positive if `skip_solid_cache?` is available here, but I think it is based on other database templates. If so, would be good to use it here too.

djmb committed:
Update railties/lib/rails/generators/app_base.rb

Co-authored-by: Jeffrey Hardy <packagethief@gmail.com>

djmb committed:
Use skip_solid_cache? in templates

Adding this highlighted that we need to add support to the change
template for deciding whether to include the solid cache configuration
when migrating.

djmb commented:
Closed in favour of https://github.com/rails/rails/pull/52790
